rust-runtime/smithy-json/src/deserialize.rs

```diff
@@ -0,0 +1,850 @@
+                b' ' | b'\t' | b'\r' | b'\n' => {
+                    self.advance();
+                }
+                _ => break,
+            }
+        }
+    }
+
+    /// Returns the top of the state stack (current state).
+    fn state(&self) -> State {
+        self.state_stack[self.state_stack.len() - 1]
+    }
+
+    /// Replaces the top of the state stack with a new `state`.
+    fn replace_state(&mut self, state: State) {
+        self.state_stack.pop();
+        self.state_stack.push(state);
+    }
+
+    /// Discards the '{' character and pushes the `ObjectFirstKeyOrEnd` state.
+    fn start_object(&mut self) -> Token {
+        let byte = self.next_byte();
+        debug_assert_eq!(byte, Some(b'{'));
+        self.state_stack.push(State::ObjectFirstKeyOrEnd);
+        Token::StartObject
+    }
+
+    /// Discards the '}' character and pops the current state.
+    fn end_object(&mut self) -> Token {
+        let (byte, state) = (self.next_byte(), self.state_stack.pop());
+        debug_assert_eq!(byte, Some(b'}'));
+        debug_assert!(
+            state == Some(State::ObjectFirstKeyOrEnd) || state == Some(State::ObjectNextKeyOrEnd)
+        );
+        Token::EndObject
+    }
+
+    /// Discards the '[' character and pushes the `ArrayFirstValueOrEnd` state.
+    fn start_array(&mut self) -> Token {
+        let byte = self.next_byte();
+        debug_assert_eq!(byte, Some(b'['));
+        self.state_stack.push(State::ArrayFirstValueOrEnd);
+        Token::StartArray
+    }
+
+    /// Discards the ']' character and pops the current state.
+    fn end_array(&mut self) -> Token {
+        let (byte, state) = (self.next_byte(), self.state_stack.pop());
+        debug_assert_eq!(byte, Some(b']'));
+        debug_assert!(
+            state == Some(State::ArrayFirstValueOrEnd) || state == Some(State::ArrayNextValueOrEnd)
+        );
+        Token::EndArray
+    }
+
+    /// Reads a JSON Unicode escape sequence (i.e., "\u1234").
+    fn read_unicode_escape(&mut self, into: &mut Vec<u8>) -> Result<(), Error> {
+        let (start, end) = (self.index, self.index + 4);
+        if end > self.input.len() {
+            return Err(self.error(UnexpectedEOS));
+        }
+
+        let codepoint_str =
+            std::str::from_utf8(&self.input[start..end]).map_err(|err| self.error(err.into()))?;
+        let codepoint = u32::from_str_radix(codepoint_str, 16)
+            .map_err(|_| self.error(ErrorReason::InvalidUnicodeEscape(codepoint_str.into())))?;
+        let codepoint = char::from_u32(codepoint)
+            .ok_or_else(|| self.error(InvalidUnicodeEscape(codepoint_str.into())))?;
+        match codepoint.len_utf8() {
+            1 => into.push(codepoint as u8),
+            _ => into.extend_from_slice(codepoint.encode_utf8(&mut [0; 4]).as_bytes()),
+        }
+        self.index = end;
+        Ok(())
+    }
+
+    /// Reads a JSON string out of the stream.
+    fn read_string(&mut self) -> Result<String, Error> {
```

Comment:
the XML parsing library I used didn't unescape in the tokenizer. That enabled the tokenizer to be allocation-free (and for input that you _know_ won't need to be escaped (like base64), you can even skip the unescape step later)

I don't think we want to be allocating for every string we read out of the input stream, especially since lots of them (eg. a key in a map) we never need to own

Reply:
This makes a lot of sense. Will refactor.